azkaban-executor/azkaban-emr-jobtype/src/main/java/uk/gov/dwp/dataworks/azkaban/services/DataProductStatusService.java

```diff
@@ -0,0 +1,120 @@
+    }
+
+    @Override
+    public void cancel() {
+        super.cancel();
+        this.dependentProductLatch.countDown();
+    }
+
+    private Optional<Boolean> completedSuccessfully(final String tableName, final Map<String, AttributeValue> item) {
+        final ScheduledExecutorService timeoutExecutor = Executors.newSingleThreadScheduledExecutor();
+        final ScheduledExecutorService dependencyCheckExecutor = Executors.newSingleThreadScheduledExecutor();
+        try {
+            final AtomicBoolean succeeded = new AtomicBoolean(false);
+            dependencyCheckExecutor.scheduleWithFixedDelay(() -> checkDependency(tableName, item, succeeded), 0,
+                    pollIntervalMilliseconds(), TimeUnit.MILLISECONDS);
+            timeoutExecutor
+                    .schedule(dependentProductLatch::countDown, pollTimeoutMilliseconds(), TimeUnit.MILLISECONDS);
+            dependentProductLatch.await();
+            return Optional.of(succeeded.get());
+        } catch (InterruptedException e) {
+            error("Failed to check dependency '" + tableName + "', item: '" + item + "'", e);
+            return Optional.empty();
+        } finally {
+            timeoutExecutor.shutdownNow();
+            dependencyCheckExecutor.shutdownNow();
+        }
+
+    }
+
+    private void checkDependency(final String tableName, final Map<String, AttributeValue> item,
+            final AtomicBoolean succeeded) {
+        try {
+            info("Checking '" + itemString(item) + "'");
+            final GetItemRequest request = new GetItemRequest().withTableName(tableName).withKey(primaryKey(item));
+            final GetItemResult result = dynamoDb.getItem(request);
+            final String status = result.getItem().get(STATUS_FIELD).getS();
+            info("Checked '" + itemString(item) + "', status is '" + status + "'.");
+            if (hasFinished(status)) {
+                info("Dependency '" + itemString(item) + "' has completed, status is '" + status + "'.");
+                succeeded.set(hasSucceeded(status));
+                dependentProductLatch.countDown();
+            }
+        } catch (Exception e) {
+            error("Failed to check item:", e);
+        }
+    }
+
+    private static boolean hasFinished(final String status) {
+        return hasSucceeded(status) || hasFailed(status);
+    }
+
+    private static boolean hasFailed(final String status) {
+        return hasStatus(FAILED_COMPLETION_STATUS, status);
+    }
+
+    private static boolean hasSucceeded(final String status) {
+        return hasStatus(SUCCESSFUL_COMPLETION_STATUS, status);
+    }
+
+    private static boolean hasStatus(final String required, final String status) {
+        return required.equalsIgnoreCase(status);
+    }
+
+    private static String itemString(final Map<String, AttributeValue> item) {
+        return item.get(CORRELATION_ID_FIELD).getS() + "/" + item.get(DATA_PRODUCT_FIELD).getS() + "/" + item
```

Comment:
Where are `CORRELATION_ID_FIELD`, `DATE_FIELD` and `DATA_PRODUCT_FIELD` set?

Reply:
They're in the MetadataService interface